File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int l = 0;
        int r = height.length;
        int lp = height[0];
        int rp = height[r-1];
        int water = 0;
        while(l < r)
        {
            if(lp < rp)
            {
                l++;
                lp = Math.max(lp, height[l]);
                water += lp - height[l];
            }
            else
            {
                r--;
                rp = Math.max(rp, height[r]);
                water += rp - height[r];
            }
        }
        return water;     
    }
}

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        i = 0
        j = len(height) - 1
        tot = 0
        h1 = height[0]
        h2 = height[j]
        while(i < j):
            if height[i] < height[j]:
                h1 = max(h1, height[i])
                if h1 - height[i] > 0:
                    tot += h1 - height[i]
                i += 1
            else:
                h2 = max(h2,height[j])
                if h2 - height[j] > 0:
                    tot += h2 - height[j]
                j -= 1


        return tot